internal/pipe/krew/krew.go

```diff
@@ -0,0 +1,325 @@
+	out, err := yaml.Marshal(data)
+	if err != nil {
+		return "", fmt.Errorf("krew: failed to marshal yaml: %w", err)
+	}
+	return string(out), nil
+}
+
+func dataFor(ctx *context.Context, cfg config.Krew, cl client.Client, artifacts []*artifact.Artifact) (Plugin, error) {
+	desc, err := tmpl.New(ctx).Apply(cfg.Description)
+	if err != nil {
+		return Plugin{}, err
+	}
+
+	shortDesc, err := tmpl.New(ctx).Apply(cfg.ShortDescription)
+	if err != nil {
+		return Plugin{}, err
+	}
+
+	result := Plugin{
+		APIVersion: apiVersion,
+		Kind:       kind,
+		Metadata: Metadata{
+			Name: cfg.Name,
+		},
+		Spec: Spec{
+			Homepage:         cfg.Homepage,
+			Version:          ctx.Version,
+			ShortDescription: shortDesc,
+			Description:      desc,
+			Caveats:          cfg.Caveats,
+		},
+	}
+
+	for _, art := range artifacts {
+		sum, err := art.Checksum("sha256")
+		if err != nil {
+			return result, err
+		}
+
+		if cfg.URLTemplate == "" {
+			url, err := cl.ReleaseURLTemplate(ctx)
+			if err != nil {
+				return result, err
+			}
+			cfg.URLTemplate = url
+		}
+		url, err := tmpl.New(ctx).WithArtifact(art, map[string]string{}).Apply(cfg.URLTemplate)
+		if err != nil {
+			return result, err
+		}
+
+		goarch := []string{art.Goarch}
+		if art.Goarch == "all" {
+			goarch = []string{"amd64", "arm64"}
+		}
+
+		for _, arch := range goarch {
+			platform := Platform{
+				URI:    url,
+				Sha256: sum,
+				Selector: Selector{
+					MatchLabels: MatchLabels{
+						Os:   art.Goos,
+						Arch: arch,
+					},
+				},
+			}
+			for _, bin := range art.ExtraOr(artifact.ExtraBinaries, []string{}).([]string) {
+				platform.Bin = bin
+				break
+			}
+			result.Spec.Platforms = append(result.Spec.Platforms, platform)
+		}
+	}
+
+	return result, nil
+}
+
+// Publish krew plugin.
+func (Pipe) Publish(ctx *context.Context) error {
+	cli, err := client.New(ctx)
+	if err != nil {
+		return err
+	}
+	return publishAll(ctx, cli)
+}
+
+func publishAll(ctx *context.Context, cli client.Client) error {
+	skips := pipe.SkipMemento{}
+	for _, plugin := range ctx.Artifacts.Filter(artifact.ByType(artifact.KrewPlugin)).List() {
+		err := doPublish(ctx, plugin, cli)
+		if err != nil && pipe.IsSkip(err) {
+			skips.Remember(err)
+			continue
+		}
+		if err != nil {
+			return err
+		}
+	}
+	return skips.Evaluate()
+}
+
+func doPublish(ctx *context.Context, plugin *artifact.Artifact, cl client.Client) error {
+	cfg := plugin.Extra[krewConfigExtra].(config.Krew)
+	var err error
+	cl, err = client.NewIfToken(ctx, cl, cfg.Index.Token)
+	if err != nil {
+		return err
+	}
+
+	if strings.TrimSpace(cfg.SkipUpload) == "true" {
+		return pipe.Skip("krews.skip_upload is set")
+	}
+
+	if strings.TrimSpace(cfg.SkipUpload) == "auto" && ctx.Semver.Prerelease != "" {
+		return pipe.Skip("prerelease detected with 'auto' upload, skipping krew publish")
+	}
+
+	repo := client.RepoFromRef(cfg.Index)
+
+	gpath := buildPluginPath(pluginsFolder, plugin.Name)
```

Comment:
Hi @caarlos0,

Can we add here maybe the https://github.com/kubernetes-sigs/krew/tree/master/cmd/validate-krew-manifest

> Ensure plugin manifests are valid YAML and pass Krew manifest validation (optionally, you can use the validate-krew-manifest tool for static analysis).


to validate the krew file? just in case?

Reply:
yeah, was working on it :)